### PR TITLE
Redirect supported product types to the new experience based on the product template associated to it

### LINKED ISSUE #43224: Redirect supported product types to the new experience based on the product template associated to it
**Prerequisites (mark completed items with an [x]):**
- [ ] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [ ] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Description**

When the product is created based on a selected product template, if the product template is not listed as part of the supported types then the user is redirected to the classic experience and it should not.

**ACs**
```[tasklist]
### Tasks
- [ ] Only redirect users to the classic experience (a) if the product does not has a product template, (b) if it has a product template but not a layout template.
```

### PR DESCRIPTION
**Pull request by mdperez86**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #43224

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`
2. Go to `Products` -> `Add new` from the left side menu.
3. Set the `Name` of the product.
4. Under the `General` tab a description `This is a standard product. Change product type` should be shown below the `Basic details` section.
5. When clicking the `Change product type` a dropdown menu should be shown listing all the available product types. 
image

9. Add some custom product types like follows. You can use a Code Snipped plugin or create a custom plugin as you want.
```php
add_action( 'init', 'register_demo_product_type', 0 );
function register_demo_product_type() {
	class WC_Product_Demo extends WC_Product {
		public function __construct( $product ) {
			parent::__construct( $product );
		}

		public function get_type() {
			return 'demo';
		}
	}
}

add_filter( 'product_type_selector', 'add_demo_product_type' );
function add_demo_product_type( $types ){
	$types[ 'demo' ] = __( 'Demo product', 'dm_product' );
	return $types; 
}

add_filter( 'woocommerce_product_editor_product_templates', 'register_product_template' );
function register_product_template( $templates ) {
	$templates[] = new Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplate(
		array(
			'id'                 => 'demo-product-template',
			'title'              => __( 'Product demo', 'woocommerce' ),
			'description'        => __( 'This is a demo product', 'woocommerce' ),
			'order'              => ( count( $templates ) + 1 ) * 10,
			'icon'               => 'bug',
			'layout_template_id' => 'simple-product',
			'product_data'       => array(
				'type' => 'demo',
			),
		)
	);
	
	return $templates;
}
```
17. With that you added support to the new product type created in point 9. And the menu now should looks like 
image

19. Clicking the `Product demo` item should save the changes, apply the `simple-product` template to the editor and keep you in the same page without redirecting to the classic experience. 
20. Try to edit the same product from the product list. 
21. You should now be redirected to the new product edition experience and not to the classic.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/RedirectionController.php [hunk 1]
```diff
@@ -62,8 +69,17 @@ protected function is_legacy_edit_screen(): bool {
...
+		$product             = $product_id ? wc_get_product( $product_id ) : null;
+		$digital_product     = $product->is_downloadable() || $product->is_virtual();
+		$product_template_id = $product->get_meta( '_product_template_id' );
+
+		if ( isset( $product_template_id ) ) {
+			foreach ( $this->product_templates as $product_template ) {
+				if ( $product_template_id === $product_template->get_id() && ! is_null( $product_template->get_layout_template_id() ) ) {
+					return true;
+				}
+			}
```
Comment (louwie17) [inline plugins/woocommerce/src/Admin/Features/ProductBlockEditor/RedirectionController.php:79]: This seems to break the redirection when a template id is already set in the meta and a user tries to redirect to a unsupported product type.
It appears that when we switch to an un-supported product type we don't remove the `_product_template_id` meta field that was previously set. Maybe we should remove it? Or run the logic below first?

**Review by louwie17 — CHANGES_REQUESTED** [review]
Overall looks good, but did notice one bug when redirecting to an unsupported template after a template id is already set:

https://github.com/woocommerce/woocommerce/assets/2240960/d56cdc66-b9a3-4607-944c-f4cd42a4f529

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/RedirectionController.php (hunk 1 above)
Comment (mdperez86) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/RedirectionController.php:79]: Done -> https://github.com/woocommerce/woocommerce/pull/43341/commits/c76b6ac3afc5f07a25823db97212c47d0e5eabb1

**Review by louwie17 — APPROVED** [review]
Thanks for the updates @mdperez86, this is testing well this time around, nice work!!

When testing I did notice an alternate issue, not related to this PR, and can be addressed as part of a separate issue. It looks like some plugins will hit validation errors when just changing the product type blindly, like product bundles in the below video, it complains about specific values when trying to save. 

Weirdly enough I only hit this error when changing the type of a brand new product, once a product has been created there was no issue with switching types. So maybe this is a product bundle plugin issue.

https://github.com/woocommerce/woocommerce/assets/2240960/a268b786-6a24-4aeb-9907-70acd0bf4a98

**Pull request merged**